src/serac/numerics/functional/dual.hpp

```diff
@@ -199,13 +199,80 @@ SERAC_HOST_DEVICE constexpr auto& operator-=(dual<gradient_type>& a, double b)
   return a;
 }
 
-/** @brief implementation of absolute value function for dual numbers */
+/**
+ * @brief Implementation of absolute value function for dual numbers
+ * @note This is not differentiable at x = 0.0
+ */
 template <typename gradient_type>
 SERAC_HOST_DEVICE auto abs(dual<gradient_type> x)
 {
   return (x.value >= 0) ? x : -x;
 }
 
+/**
+ * @brief Implementation of max for dual numbers
+ * @note This is not differentiable at a == b
+ */
+template <typename gradient_type>
+SERAC_HOST_DEVICE auto max(dual<gradient_type> a, double b)
+{
+  auto b_dual = dual{b, 0.0};
+  return (a > b_dual) ? a : b_dual;
+}
+
+/**
+ * @brief Implementation of max for dual numbers
+ * @note This is not differentiable at a == b
```

Comment:
this note is good, maybe we could also say what it does in the case of equality (i.e. just returns the derivative of the second argument)

Another option might be to average the derivatives of the two operands when equal (this is probably overthinking things)